tools/scoring/score.py

```diff
@@ -568,28 +606,124 @@ def project(normal,other):
     other_different = different.loc['other']
     return other_different
 
+# Do not consider solver variants for determining if a division is
+# competitive.
+# Uses solver_str(solver) to get the base version of solver.
+def is_division_competitive(solvers):
+    return len(set([solver_str(x) for x in solvers])) > 1
+
+
+# Biggest Lead Ranking.
+#
+# Computes the new global ranking based on the distance between the winner of a
+# division and the second solver in a division as defined in secion 7.3.1 of
+# the SMT-COMP'19 rules.
+#
+# Note: The function prints a list of sorted tuples starting with the first
+#       place (winner).
+#
+def biggest_lead_ranking(data):
+    start = time.time() if g_args.show_timestamps else None
+
+    data = group_and_rank_solver(data)
+    data = data[data.competitive == True]
+    scores = []
+    # TODO: use for (year, division) groupby(level ...)
+    for year, ydata in data.groupby('year'):
+        for division, div_data in ydata.groupby('division'):
+
+            # Skip non-competitive divisions
+            if not is_division_competitive(div_data.solver.unique()):
+                continue
+
+            assert len(div_data) >= 2
+            first = div_data.iloc[0]
+            second = div_data.iloc[1]
+            score = ((1 + first.correct) / (1 + second.correct))
+            scores.append(
+                (score, first.solver, second.solver, division))
+
+    scores_sorted = sorted(scores, reverse=True)
+    print('Biggest Lead Ranking (Score, 1st Solver, 2nd Solver, Division)')
+    for s in scores_sorted:
+        print(s)
+
+    if g_args.show_timestamps:
+        log('time biggest_lead_ranking: {}'.format(time.time() - start))
+
+
+# Largest Contribution Ranking.
+#
+# Compute the correctly solved score for the virtual best solver for a given
+# division if the results of 'solver' are excluded. This function corresponds
+# to function vbss(D,S) as defined in section 7.3.2 of the SMT-COMP'19 rules.
+#
+def vbss(division_data, solver):
+
+    # For VBS we only consider correctly solved benchmarks
+    data = division_data[(division_data.solver != solver)
+                         & (division_data.correct > 0)
+                         & (division_data.error == 0)]
+
+    sort_columns = ['benchmark', 'score_correct', 'wallclock_time']
+    sort_asc = [True, False, True]
+
+    # Get job pair with the highest correctly solved score, which was solved
+    # the fastest.
+    data_vbs = data.sort_values(
+                by=sort_columns, ascending=sort_asc).groupby(
+                        'benchmark', as_index=False).first()
+    assert len(data_vbs) == len(data.benchmark.unique())
+    return data_vbs.score_correct.sum()
```

Comment:
Is this the correct thing to return? In the rules document, we take the (score * time) for the 'best' solution to each benchmark

Reply:
What is the rational behind `score * time`?

Reply:
For selecting the fastest solver for the VBS, don't you want the highest score with the fastest CPU time?

Reply:
Yes, I realised this earlier today e.g. that we're producing a number to maximise but then putting the lowest time in there... 

The score_correct is just going to be '1' per benchmark isn't it (for this year where we don't use families) - I thought I'd checked this with the previous bit of the code but I may have gotten this wrong.

The idea of including the time is that if all solvers solve exactly the same benchmarks we still see the benefit of adding or removing solvers, which we wouldn't do if we just take the number of benchmarks solved.

I think what I wanted to have was (score/time) instead of (score * time) ... although we might need to be careful of small numbers.

Reply:
@selig the problem is that this doesn't work for the incremental track. In the incremental track we would punish solvers that answer more check-sat calls with that formula. For unsat core it should be fine (because there it is reduction rather than number of sat calls).

Reply:
@selig Here is a new proposal that @aniemetz and I discussed offline.
The current `vbss(D,S)` scoring scheme in the rules is flawed. As mentioned in the rules document, for the single-query track `vbss(D,S)` essentially returns the CPU time. Removing a solver from `S` can only make the virtual best solver slower, i.e., `vbss(D, S-s) >= vbss(D, S)`. If `vbss(D,S-s) > vbss(D,S)`, the formula in the rules computes values < 0. Additionally, since only solved benchmarks are considered (`n^s_b > 0`) it can happen that by removing a solver with uniquely solved instances from `S`, the number of considered benchmark in the division changes. Resource outs are not considered at all.

Here is a new proposal for the VBSS score:

Instead of returning a single score, `vbss(D,S)` returns a tuple `(correct_score, cpu_time)`, which corresponds to the sum of correct scores and CPU time of the VBS given the set of solvers `S`  for **all** benchmarks in the division, where `vbss_c(D,S)` corresponds to the `correct_score` of the tuple and `vbss_t(D,S)` corresponds to the `cpu_time` of the tuple. Note that for this scoring scheme, solvers need to be penalized if they are not able to solve a benchmark (no matter if  they report unknown or memout), i.e., the CPU time is set to the time limit.

For each solver we now compute two scores:

1) `correct_score_impact = 1 - vbss_c(D,S-s) / vbss_c(D,S)`
2) `cpu_time_impact = 1 - vbss_t(D,S) / vbss_t(D,S-s)`

where in both cases, a higher score means a higher impact on the VBS (with score 0 meaning no impact at all).

For each division solvers are then ranked by the tuple `(correct_score_impact, cpu_time_impact)`.